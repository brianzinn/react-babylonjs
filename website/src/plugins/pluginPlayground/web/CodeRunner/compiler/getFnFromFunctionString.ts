import getImport from '_playground_virtual_modules'

const GET_IMPORT_FN = '__get_import'

type ExportObject = 'module' | 'exports'
type ExportProperty = 'exports' | 'default'

export function getComponentFnFromCodeString(code: string) {
  /**
   * In CommonJS default export is either of the two forms:
   * - `module.exports = Component`
   * - `exports.default = Component`
   *
   * We will plug in our `exportObject` into the function
   * to get the default exported componentFn assigned to it.
   */
  const exportObject: {
    [Prop in ExportProperty]?: React.FC
  } = {}

  const [OBJECT_NAME, ASSIGN_TO_PROP] = resolveExportType(code)

  const fnArgNames = [GET_IMPORT_FN, OBJECT_NAME] as const
  const fnCode = replaceRequireWithGetImport(code)

  const func = new Function(...fnArgNames, fnCode) as (
    getImportFn: typeof getImport,
    exportsObj: typeof exportObject
  ) => void

  /**
   * After this call:
   * - `getImport` would resolve external module imports
   * - `exportObject` would be `{ [ExportProperty]: componentFn }`
   *
   * Thus, we can grab the componentFn from `exportObject`
   */
  func(getImport, exportObject)

  const componentFn = exportObject[ASSIGN_TO_PROP]

  return componentFn
}

function resolveExportType(code: string): [ExportObject, ExportProperty] {
  if (code.includes('module.exports')) {
    return ['module', 'exports']
  } else if (code.includes('exports.default')) {
    return ['exports', 'default']
  } else {
    throw new Error(`Missing default export in the file`)
  }
}

/**
 * Replaces `require('module')` with `__get_import('module')`
 * to later call `getImport` that resolves external modules.
 *
 * This might better be done with a proper AST traversal,
 * but let's just say we are following the KISS principle here
 * (if we can say that after all of the above)
 */
function replaceRequireWithGetImport(code: string) {
  let fnString = code.replaceAll('require(', `${GET_IMPORT_FN}(`)

  if (!fnString.includes('const React =')) {
    fnString = `const React = ${GET_IMPORT_FN}('react');\n${fnString}`
  }

  return fnString
}
