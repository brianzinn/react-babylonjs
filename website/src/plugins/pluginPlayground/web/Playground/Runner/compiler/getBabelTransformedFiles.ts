import type babel from '@babel/standalone'
import { Files } from './getComponentFromCode'

type Babel = typeof babel

// @babel/standalone is loaded with html script tag
// see builderConfig.html.tags in pluginPlayground
declare global {
  interface Window {
    Babel: Babel
  }
}

export function getBabelTransformedFiles(files: Files) {
  return Object.keys(files).reduce<Files>((acc, fileName) => {
    const code = files[fileName]

    const transformResult = window.Babel.transform(code, {
      presets: [
        [window.Babel.availablePresets.react, { pure: false }],
        [window.Babel.availablePresets.typescript, { allExtensions: true, isTSX: true }],
      ],
    })

    const transformedCode = transformResult?.code

    if (transformedCode) {
      acc[fileName] = transformedCode
    }

    return acc
  }, {})
}
