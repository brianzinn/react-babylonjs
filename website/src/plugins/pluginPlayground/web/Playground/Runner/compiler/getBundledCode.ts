import { Plugin, rollup } from '@rollup/browser'
import { prepareFileNameWithExt } from '@/src/plugins/pluginPlayground/cli/utils/getImport'
import { Files } from './getComponentFromCode'

export const getBundledCode = async ({
  entryFilename,
  files,
}: {
  entryFilename?: string
  files: Files
}) => {
  const bundle = await rollup({
    input: entryFilename,
    plugins: [
      loaderPlugin(files),
      {
        name: 'test-transform',
        transform(code, id) {
          // console.log({ code, id })
          return code
        },
      },
    ],
    // regexp to match non-local imports, ie. it does not start with . or /
    external: /^[^(.|/)]/,
  })

  const { output } = await bundle.generate({
    compact: false,
    strict: false,
    format: 'commonjs',
    generatedCode: 'es2015',
    plugins: [],
  })

  const bundledCodeString = output[0].code
  // console.log({ bundledCodeString })

  return bundledCodeString
}

function loaderPlugin(files: Files): Plugin {
  return {
    name: 'loader',
    resolveId(source, importer, options) {
      if (files.hasOwnProperty(source)) {
        return source
      }

      if (source.startsWith('.')) {
        const fileName = `/${prepareFileNameWithExt(source)}`

        if (files.hasOwnProperty(fileName)) {
          return fileName
        }
      }

      return source
    },
    load(id) {
      if (files.hasOwnProperty(id)) {
        return files[id]
      }
    },
  }
}
