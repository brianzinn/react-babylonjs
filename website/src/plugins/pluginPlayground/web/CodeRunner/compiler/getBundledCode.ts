import { Plugin, rollup } from '@rollup/browser'
import { ENTRY_FILE_NAME } from '../../../shared/constants'
import { isRelativeImport, prepareFileNameWithExt } from '../../../shared/pathHelpers'
import { Files } from './compileComponentFromFiles'

export const getBundledCode = async ({ files }: { files: Files }) => {
  const entryFilename = Object.keys(files).find((name) => name.includes(ENTRY_FILE_NAME))

  const bundle = await rollup({
    input: entryFilename,
    plugins: [loaderPlugin(files)],
    external: (source) => {
      const isLocal = isRelativeImport(source) || files[source]

      return !isLocal
    },
  })

  const { output } = await bundle.generate({
    compact: false,
    strict: false,
    exports: 'named',
    format: 'commonjs',
    generatedCode: 'es2015',
    plugins: [],
  })

  const bundledCode = output[0].code

  return bundledCode
}

// Based off of https://rollupjs.org/faqs/#how-do-i-run-rollup-itself-in-a-browser
// Resolve and load modules to be bundled
function loaderPlugin(files: Files): Plugin {
  return {
    name: 'in-memory-loader',
    resolveId(source) {
      if (files.hasOwnProperty(source)) {
        return source
      }

      // resolve file name from the relative import
      if (isRelativeImport(source)) {
        const fileName = `/${prepareFileNameWithExt(source)}`

        if (files.hasOwnProperty(fileName)) {
          return fileName
        }
      }

      return null
    },
    load(fileName) {
      // use fileName resolved above to load its code for bundling
      if (files.hasOwnProperty(fileName)) {
        return files[fileName]
      }
    },
  }
}
