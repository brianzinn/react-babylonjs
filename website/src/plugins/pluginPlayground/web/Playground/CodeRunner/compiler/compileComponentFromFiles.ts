import { getBundledCode } from './getBundledCode'
import { getBabelTransformedFiles } from './getBabelTransformedFiles'
import { getComponentFnFromCodeString } from './getFnFromFunctionString'

export type Files = Record<string, string>

export const compileComponentFromFiles = async (files: Files) => {
  if (!window.Babel) return

  // Rollup can't handle TSX, and its typescript plugin doesn't work in the browser.
  // Using @babel/standalone to transform TSX into JS
  const babelTransformedFiles = getBabelTransformedFiles(files)

  // Bundle files into a single chunk
  const bundledCode = await getBundledCode({
    files: babelTransformedFiles,
  })

  return getComponentFnFromCodeString(bundledCode)
}
