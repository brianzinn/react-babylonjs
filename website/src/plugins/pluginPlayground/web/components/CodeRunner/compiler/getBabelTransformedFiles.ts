import type babel from '@babel/standalone'
import { TransformOptions } from '@babel/core'
import { Language } from '@pluginPlayground/shared/constants'
import { Files } from './compileComponentFromFiles'

type Babel = typeof babel

// @babel/standalone is loaded with html script tag
// see builderConfig.html.tags in pluginPlayground
declare global {
  interface Window {
    Babel: Babel
  }
}

type GetBabelTransformedFiles = {
  files: Files
  language: Language
}

export function getBabelTransformedFiles({ files, language }: GetBabelTransformedFiles) {
  const { availablePresets, transform } = window.Babel

  const presets: TransformOptions['presets'] = [[availablePresets.react, { pure: false }]]

  if (language === Language.tsx) {
    presets.push([availablePresets.typescript, { allExtensions: true, isTSX: true }])
  }

  return Object.keys(files).reduce<Files>((acc, fileName) => {
    const code = files[fileName]

    const fileResult = transform(code, { presets })

    if (fileResult?.code) {
      acc[fileName] = fileResult.code
    }

    return acc
  }, {})
}
