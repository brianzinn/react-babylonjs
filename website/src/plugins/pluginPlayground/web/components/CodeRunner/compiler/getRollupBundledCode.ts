import { Plugin, rollup } from '@rollup/browser'
import { Language, ENTRY_FILE_BASE } from '@pluginPlayground/shared/constants'
import {
  isRelativeImport,
  prepareFileNameWithExt,
} from '@/src/plugins/pluginPlayground/shared/pathHelpers'
import { Files } from './compileComponentFromFiles'

type GetBundledCode = {
  files: Files
  language: Language
}

export const getRollupBundledCode = async ({ files, language }: GetBundledCode) => {
  const entryFilename = Object.keys(files).find((name) => name.includes(ENTRY_FILE_BASE))

  const bundle = await rollup({
    input: entryFilename,
    plugins: [loaderPlugin(files, language)],
    external: (source) => {
      const isLocal = isRelativeImport(source) || files[source]

      return !isLocal
    },
  })

  const { output } = await bundle.generate({
    compact: false,
    strict: false,
    exports: 'named',
    format: 'commonjs',
    generatedCode: 'es2015',
    plugins: [],
  })

  const bundledCode = output[0].code

  return bundledCode
}

// Based off of https://rollupjs.org/faqs/#how-do-i-run-rollup-itself-in-a-browser
// Resolve and load modules to be bundled
function loaderPlugin(files: Files, language: Language): Plugin {
  return {
    name: 'in-memory-loader',
    resolveId(source) {
      if (files.hasOwnProperty(source)) {
        return source
      }

      // resolve file name from the relative import
      if (isRelativeImport(source)) {
        const fileName = `/${prepareFileNameWithExt(source, language)}`

        if (files.hasOwnProperty(fileName)) {
          return fileName
        }
      }

      return null
    },
    load(fileName) {
      // use fileName resolved above to load its code for bundling
      if (files.hasOwnProperty(fileName)) {
        return files[fileName]
      }
    },
  }
}
