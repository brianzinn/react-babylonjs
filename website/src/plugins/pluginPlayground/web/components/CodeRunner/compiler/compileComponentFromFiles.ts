import { Language } from '@pluginPlayground/shared/constants'
import { getRollupBundledCode } from './getRollupBundledCode'
import { getBabelTransformedFiles } from './getBabelTransformedFiles'
import { getComponentFnFromCodeString } from './getFnFromFunctionString'

export type Files = Record<string, string>

export const compileComponentFromFiles = async (files: Files, language: Language) => {
  if (!window.Babel) return

  // Rollup requires plugins to handle JSX/TSX,
  // but they don't work in the browser.
  // Using @babel/standalone to transform JSX/TSX into JS
  const babelTransformedFiles = getBabelTransformedFiles({ files, language })

  // Bundle files into a single chunk
  const bundledCode = await getRollupBundledCode({
    language,
    files: babelTransformedFiles,
  })

  return getComponentFnFromCodeString(bundledCode)
}
