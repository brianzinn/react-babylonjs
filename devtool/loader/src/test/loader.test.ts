/**
 * @jest-environment node
 */
import compiler from './compiler'

test('Gets devtool result', async () => {
  const stats = await compiler('../codesandbox-templates/ts/src/App.tsx')
  const output = stats?.toJson({ source: true }).modules?.[0].source?.toString() || ''
  expect(output).toMatchSnapshot()
})
