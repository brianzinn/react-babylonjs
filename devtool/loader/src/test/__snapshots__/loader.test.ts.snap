// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gets devtool result 1`] = `"module.exports = {\\"ts\\":\\"// App.tsx\\\\n\\\\n{\\\\n  /**\\\\n   ********* WARNING ********\\\\n   * THIS CODE WAS LAUNCHED FROM LOCALHOST.\\\\n   * LOCAL PACKAGES MAY NOT MATCH THE DEPENDENCIES\\\\n   * USED IN THIS SANDBOX.\\\\n   *\\\\n   * TEST SANDBOX FOR BACKWARD COMPATIBLITY IF DESIRED.\\\\n   ********* WARNING ********\\\\n   */\\\\n}\\\\n\\\\nimport { FC } from 'react'\\\\nimport './styles.css'\\\\n\\\\nconst App: FC = () => {\\\\n  return (\\\\n    <div className=\\\\\\"App\\\\\\">\\\\n      <h1>Hello CodeSandbox</h1>\\\\n      <h2>Start editing to see some magic happen!</h2>\\\\n    </div>\\\\n  )\\\\n}\\\\n\\",\\"js\\":\\"// App.jsx\\\\n\\\\n{\\\\n  /**\\\\n   ********* WARNING ********\\\\n   * THIS CODE WAS LAUNCHED FROM LOCALHOST.\\\\n   * LOCAL PACKAGES MAY NOT MATCH THE DEPENDENCIES\\\\n   * USED IN THIS SANDBOX.\\\\n   *\\\\n   * TEST SANDBOX FOR BACKWARD COMPATIBLITY IF DESIRED.\\\\n   ********* WARNING ********\\\\n   */\\\\n}\\\\nimport './styles.css'\\\\nconst App = () => {\\\\n  return (\\\\n    <div className=\\\\\\"App\\\\\\">\\\\n      <h1>Hello CodeSandbox</h1>\\\\n      <h2>Start editing to see some magic happen!</h2>\\\\n    </div>\\\\n  )\\\\n}\\\\n\\",\\"tsUrl\\":\\"https://codesandbox.io/api/v1/sandboxes/define?parameters=N4IgZglgNgpgziAXKAdABwE4wC7YjDDAYyVCIHsA7bGapEYAHUoAIXGRMJqB1CAE2wALDohYAOAAwAaZmw7YAhgCM-gkUhYAmWa3YgArnBgAVFXFEswiqMd3yQxgLYRL12zHv643AOawARQNyGktsDANPOX1wxWg_AGFyJydFSw5KKhgOLw5lDEUiAGscAGU0Qr8wiKi9DgArOAAPACEC4rLFJxgAGW5szXc7aI4wKAMBNxtjZgBfZhBZ6U5CosVfGBRGqlIQCmpabHomOpBKLoGxDgp-eEVKfmVyJoBaLELsF-xmnJGQADcCD4qOkQABGFCSSG_U63OBEDAQNB4EGaDgwhwlACeAHdyBh-BZNABtAC6uRAqW4oLgxAA9Nxbk0UN8mhj9Lc0LRbpQiPgiWITmwHAABZQqLFQKiNOkULCggAsKC0KDB7NF4uUkulcDpvgmiuVqvV-k12sojReQnI5CKoKhEMkJo47yI2FBYIA7JDlc6QK7PmapRaBfoAMw-sNqkBeBwBl78ZIe71QrR--PwxHI0McJVQsN-7BYrkvMDwd1okAqgCc0JA0SWf1u_wAIjAuQ9aHz4JYhcKOJgcHgCKCVQBWY0x6IOQe4fAYF5oca-bgvfG-e4QABeMBeECcaHx31HKAjCr9IqLXN1AeTPq0TqnelFV_gdPjiacd6h1cLxfgCJIhWVwgEqSppvWeiNqcmZAaGfYOHASgYMB-gZoB2YsEhigoX6ygTFA_CguhWbfCw-HQERT7CjE5bETAHwvLBmE0EhLAvC8tD_AAvI0n5-jA9QwG69GMcxZGCcJFYNhS-TkDixgYHAUAQEhljEtO-gAHypgApH6mTYCwtyKFRsb6IZLD4CwAA83EsGC0bmRkIQsOQaAAPouJQEAsDYUDonopJzAsSwgN8-yQL4WxwDsyB7FQNB0IgDB_NwRDjLc6kcCgdK0kQdIAFSFUVHDkn8FAHtABAAPLIhAVDwZpHBIYiImaOEkTOSA8AALLkPwBiwAAkgcGDudUXXNSAKnKNlID8TG-jwA-YJjmV3WNGylbxltgVsPMlCHYsywoDcdwPE8TR0niGBFHAFREJs2yULs-xJUcKV9i68CHpQPiAouWD_PgOK9tN_XKNV6lhg-0gsAAbJ6Y7lc--hmMosCocSYKSFoCrw56CMo91bZwEU2ATSSYIKpIMgSHTqM0RwZMU-5bAABItupYLVnDDmSFIwVQSFr1hflDKdsyrJvYlhz0Puh4oSwwAsFgnYYCwsxWONTgsAA5B-yT68witHiwACCaBoDryQG7lVtoCbPkHub-u5UhkrwGdcBwM7zD7Gx40hAAorA3TUCw9mJkQBgR9gKAbNgYcwPHLRYsN_AABSGza2D6wAlMwzDq7cGBZzZjssHSWnw8Hyfh4cRdi8sEuOyyzSywcyUgHSdKW9bHdspQzB9kVhXTsVU_TywPAWwASgAcsNi8AOIsNP0-TywJgc8NpQsAkNUtiHs8WwfPQWwAqovCQcyHLYsAAYvPNW9SwPQ1QkFs9BzNWlCYFA29P7fx6CwAAChbBIABpC2q8Q4H16hbAAmiwReNUTAsCQSYO-O974sBPuAkOi8T632Ggg7eV9SgPxYCvPB-8WClAtiQloNUAAaQDnwTy4TvBBmCmEsPYc_Gq88WAtCgdAue89H5H16pAkww0Wg9GGiYVBw0n4EIQcNeeD9OHCk3jPKRy814bwMdw_RdJRam1dsrVWT8Eha1tnrQ2DE3TOzNsrd2eUiywDgD7P2xdeSNSMo7MQ9io4sCzgXKOWkVbRCwNgAwGBWBZ00jZfgEB_gsAyooX2i8LjcQ4I7DgWlNJsBskIMEWkOYwCgFKQ-A0YClHuI8Z4Nk6SVNKWjcpQgtBaVKMhIyMAMl4EoL4FglMsIwBgFhZIMzUgriICwIQihra0AAITtN6V0mi7SMn_B2SwZux1xb0k9r4_xXcPr0CIPcf4uS4l6AyQ9KAigsRiDGDAJoABuaInymhiDBL8vQOIBDCEBXTXSwK2BCBgBAXwQhsAQskFCqxlAUBVz7GARKpYujQHeVhe4cAmIEAgGAaFEyvmfBsPCygYgnpjWBSc2YswgA&query=template%3Ddefault%26fontsize%3D14px%26hidenavigation%3D0%26theme%3Ddark%26extends%3Dfile%253A%252FVolumes%252FCode%252Frepos%252Fharvest%252Freact-babylonjs%252Fpackages%252Fdevtool-loader%252Fdist%252F..%252Fsrc%252Fcodesandbox-templates%252Fts%26module%3D%252Fsrc%252FApp.tsx\\",\\"jsUrl\\":\\"https://codesandbox.io/api/v1/sandboxes/define?parameters=N4IgZglgNgpgziAXKAdABwE4wC7YjDDAYyVCIHsA7bGapEYAHUoAIXGRMJqB1CAE2wALDohYAOAAwAaZmw7YAhgCM-gkUhYAmWa3YgArnBgAVFXFEswiqMd3yQxgLYRL12zHv643AOawARQNyGktsDANPOX1wxWg_AGFyJydFSw5KKhgOLw5lDEUiAGscAGU0Qr8wiKi9DgArOAAPACEC4rLFJxgAGW5szXc7aI4wKAMBNxtjZgBfZhBZ6U5CosVfGBRGqlIQCmpabHomOpBKLoGxDgp-eEVKfmVyJoBaLELsF8amnJGQADcCD4qOkQABGFCSSG_U63OBEDAQNB4EGaDgwhwlACeAHdyBh-BZNABtAC6uRAqW4oLgxAA9Nxbk0UNhmhj9Lc0LRbpQiPgiWITmwHAABZQqLFQKiNOkULCggAsKC0KDB7NF4uUkulcDpvgmiuVqvV-k12sojReQnI5CKoKhEMkJo47yI2FBYIA7JDlc6QK7PmapRaBfoAMw-sNqkBeBwBl78ZIe71QrR--PwxHI0McJVQsPovRLP63f4AERgXIetD58EsQuFHEwODwBFBKoArMaY9EHM3cPgMC80ONfNwXvjfPcIAAvGAvCBOND41ntlARhV-7BYrmZpHutEgJVKtMgaLF0577P13veJQYA9Xf0wD4vK-slhwe8H2OmiZQfhQQzBF9zgFhlH_QCez0BwaC_ICXzdN8QOzFg4OwFgXheWh_gAXkaRMnD9GB6hgN0ENfd8wJIsiD3PCl8nIHFjAwOAoAgeCSVvDgAD5UwAUj9TIMNuRQoN_DIQhYfAWAAHlwlgwWjCSzik8g0AAfRcSgIBYGwoELNhSTmBYlhAFAbjuB4niaOk8QwIo4AqIhNm2Shdn2Gg6EQBg_iwJyqB8QFhywf58BxG8YP0ABZchlGgS4WGJMMtBkFgADZPQ7cluJAMxlFgR8krBSQtAVaQWE9dLspUis4CKbB1MsYkwQVSQ0qkSQcqijg6oa9S2AACTLZqwQATlSiqSqkYyixM9yzNpIg6QAQTQNAtmaDyqC8o4fLpOkWDWjbvmYZgGzpAAqS7e2uu77pYHgVoAJQAOQASVegBxFh7vu26WBMQb3tKFgEgAeTLABRR6VtBnoVoAVVehJBqhssWAAMWe8HopYHpwYSFaekG8HShMFAAYJomehYAAFFaEgAaRWr6odB6KVoATRYV7wZMFhOZMVHAbRlhobpqHXuhlH3vZgHEdKdGWE-0WQZYUoVullpwYADUpmCbsNwH2YFzXtb1rHweelgWkZpmnuejGIeihmTHeloenekwefezHxfZ97nvRg3hT-h7HY-77fvDo2w7pebF2XB8WAAchQOkv0leALLgOBU-YfYvyO9aWAUgAKABKMueJYBssGwAwMFYcvb1k_gIH-FgiCgRQ89ei5cI4Y7eNvNhZKEMEeMGmAoClMHyFuUp7keZ5ZLpSeeLHuShC0HjSm_FgYA7vBKF8NDyE_GAYE_ZIb9SMciBYIRFHW2gAEJ193reovXjv_h_mwSu81FjLCWpnbcsA4C5wQMgPYO1Dj0CIPcf4fc67RA7k5XuWIxBjBgE0AA3NEPBTQxBgiIXoHEAhhBkPavxChbAhAwAgL4IQ2BaGSHofNZgKBjroL0GAHaLxrAuCgDgz89w4BvgIBAMADC0L4M-DYFhlAxAuQOBgCh8wFpgPpIyfBm0mjbQON5EAScVx1xYFgasGAWCzCsBgZIad4yEQLjpJcFi-FgEcU4NOGdjpuPMSndOEDs7QKIHnNxhdAoYUcSEKGsBujUDLiwRMRADBJOwCgDY2AEkwEyS0LE71-Dl1TnE7AqdgGUGYNY24GBy6yT4XSHiFVyl5MyVUxYsxZhAA&query=template%3Ddefault%26fontsize%3D14px%26hidenavigation%3D0%26theme%3Ddark%26extends%3Dfile%253A%252FVolumes%252FCode%252Frepos%252Fharvest%252Freact-babylonjs%252Fpackages%252Fdevtool-loader%252Fdist%252F..%252Fsrc%252Fcodesandbox-templates%252Fjs%26module%3D%252Fsrc%252FApp.jsx\\"};"`;
